main.go

```diff
@@ -147,7 +176,97 @@ func (fp *Firehose) handleMessage(ctx context.Context, message *hedwig.Message)
 	return err
 }
 
-func (fp *Firehose) RunLeader() {
+func (fp *Firehose) moveFilesToOutputBucket(ctx context.Context, mtmv hedwig.MessageTypeMajorVersion, currTime time.Time) error {
+	// read from staging
+	filePathPrefix := fmt.Sprintf("%s/%s", mtmv.MessageType, fmt.Sprint(mtmv.MajorVersion))
+	fileNames, err := fp.storageBackend.ListFilesPrefix(ctx, fp.processSettings.StagingBucket, filePathPrefix)
+	if err != nil {
+		return err
+	}
+	var msgs byTimestamp
+	for _, fileName := range fileNames {
+		r, err := fp.storageBackend.CreateReader(ctx, fp.processSettings.StagingBucket, fileName)
+		defer r.Close()
+		if err != nil {
+			return err
+		}
+		res, err := fp.hedwigFirehose.Deserialize(r)
+		if err != nil {
+			return err
+		}
+		for _, r := range res {
+			msgs = append(msgs, &r)
+		}
+	}
+	if len(msgs) > 0 {
+		// sort by timestamp
+		sort.Sort(msgs)
+		uploadLocation := fmt.Sprintf("%s/%s/%s/%s-%s-%s.gz", mtmv.MessageType, fmt.Sprint(mtmv.MajorVersion), currTime.Format("2006/1/2"), mtmv.MessageType, fmt.Sprint(mtmv.MajorVersion), fmt.Sprint(currTime.Unix()))
+		r, err := fp.storageBackend.CreateWriter(ctx, fp.processSettings.OutputBucket, uploadLocation)
+		if err != nil {
+			return err
+		}
+		for _, msg := range msgs {
+			payload, err := fp.hedwigFirehose.Serialize(msg)
+			if err != nil {
+				return err
+			}
+			_, err = r.Write(payload)
+			if err != nil {
+				return err
+			}
+		}
+		err = r.Close()
+		if err != nil {
+			return err
+		}
+		// delete files when written to final output path
+		for _, fileName := range fileNames {
+			// ignore errors when deleting, picked up again on next run
+			_ = fp.storageBackend.DeleteFile(ctx, fp.processSettings.StagingBucket, fileName)
+		}
+	}
+	return nil
+}
+
+func (fp *Firehose) RunLeader(ctx context.Context) {
+	currentTime := fp.clock.Now()
+	timerCh := time.After(time.Duration(fp.processSettings.ScrapeInterval) * time.Second)
+	// go through all msgs and write to msgtype folder
+	for {
+		select {
+		case <-timerCh:
+			wg := &sync.WaitGroup{}
+			for mtmv, _ := range fp.registry {
+				wg.Add(1)
+				go func(mtmv hedwig.MessageTypeMajorVersion) {
+					defer wg.Done()
+					err := fp.moveFilesToOutputBucket(ctx, mtmv, currentTime)
+					// just logs errors but will retry on next run of leader
+					if err != nil {
+						fmt.Println(err.Error())
+						fp.logger.Error(ctx, err, "moving files failed",
+							"messageType", mtmv.MessageType,
+							"messageVersion", fmt.Sprint(mtmv.MajorVersion),
+						)
+					}
+				}(mtmv)
+			}
+			wg.Wait()
+			// restart scrape interval and run leader again
+			go fp.RunLeader(ctx)
```

Comment:
I guess reading this again makes it read odd, let's just add another level of infinite loop and wrap everything in that instead of this re-spawning of goroutines for both leader and follower.

Reply:
makes sense just have it repeat over and over again